### ISSUE TITLE
Standalone authenticator does not respect the given custom port while renewing

### ISSUE THREAD

**Issue by fdellwing** [certbot/certbot#6527]
## My operating system is (include version):
`Ubuntu 16.04.5 LTS`
`Linux proxy 4.4.0-135-generic #161-Ubuntu SMP Mon Aug 27 10:45:01 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux`

## I installed Certbot with (certbot-auto, OS package manager, pip, etc):
`sudo apt install certbot`

## I ran this command and it produced this output:
`sudo certbot renew`
```
Processing
/etc/letsencrypt/renewal/www.saarlaendisches-krankenhaus-verzeichnis.de.conf
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Cert is due for renewal, auto-renewing...
Plugins selected: Authenticator standalone, Installer None
Renewing an existing certificate
Performing the following challenges:
tls-sni-01 challenge for krsaar.live.promato.de
tls-sni-01 challenge for saarlaendisches-krankenhaus-verzeichnis.de
tls-sni-01 challenge for www.saarlaendisches-krankenhaus-verzeichnis.de
Cleaning up challenges
Attempting to renew cert (www.saarlaendisches-krankenhaus-verzeichnis.de) from /etc/letsencrypt/renewal/www.saarlaendisches-krankenhaus-verzeichnis.de.conf produced an unexpected error: Problem binding to port 443: Could not bind to IPv4 or IPv6.. Skipping.
```

## Certbot's behavior differed from what I expected because:
cert did not get renewed

###### Logs are stored in `/var/log/letsencrypt` by default. Feel free to redact domains, e-mail and IP addresses as you see fit.
```certbot
2018-11-21 11:52:47,095:INFO:certbot.auth_handler:Performing the following challenges:
2018-11-21 11:52:47,096:INFO:certbot.auth_handler:tls-sni-01 challenge for krsaar.live.promato.de
2018-11-21 11:52:47,096:INFO:certbot.auth_handler:tls-sni-01 challenge for saarlaendisches-krankenhaus-verzeichnis.de
2018-11-21 11:52:47,097:INFO:certbot.auth_handler:tls-sni-01 challenge for www.saarlaendisches-krankenhaus-verzeichnis.de
2018-11-21 11:52:47,098:DEBUG:acme.standalone:Failed to bind to :443 using IPv6
2018-11-21 11:52:47,098:DEBUG:acme.standalone:Failed to bind to :443 using IPv4
2018-11-21 11:52:47,099:DEBUG:certbot.error_handler:Encountered exception:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/certbot/plugins/standalone.py", line 73, in run
    address, self.certs)  # type: acme_standalone.BaseDualNetworkedServers
  File "/usr/lib/python3/dist-packages/acme/standalone.py", line 147, in __init__
    BaseDualNetworkedServers.__init__(self, TLSSNI01Server, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/acme/standalone.py", line 108, in __init__
    raise socket.error("Could not bind to IPv4 or IPv6.")
OSError: Could not bind to IPv4 or IPv6.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/certbot/auth_handler.py", line 75, in handle_authorizations
    resp = self._solve_challenges(aauthzrs)
  File "/usr/lib/python3/dist-packages/certbot/auth_handler.py", line 126, in _solve_challenges
    resp = self.auth.perform(all_achalls)
  File "/usr/lib/python3/dist-packages/certbot/plugins/standalone.py", line 234, in perform
    return [self._try_perform_single(achall) for achall in achalls]
  File "/usr/lib/python3/dist-packages/certbot/plugins/standalone.py", line 234, in <listcomp>
    return [self._try_perform_single(achall) for achall in achalls]
  File "/usr/lib/python3/dist-packages/certbot/plugins/standalone.py", line 241, in _try_perform_single
    _handle_perform_error(error)
  File "/usr/lib/python3/dist-packages/certbot/plugins/standalone.py", line 239, in _try_perform_single
    return self._perform_single(achall)
  File "/usr/lib/python3/dist-packages/certbot/plugins/standalone.py", line 247, in _perform_single
    servers, response = self._perform_tls_sni_01(achall)
  File "/usr/lib/python3/dist-packages/certbot/plugins/standalone.py", line 264, in _perform_tls_sni_01
    servers = self.servers.run(port, challenges.TLSSNI01, listenaddr=addr)
  File "/usr/lib/python3/dist-packages/certbot/plugins/standalone.py", line 78, in run
    raise errors.StandaloneBindError(error, port)
certbot.errors.StandaloneBindError: Problem binding to port 443: Could not bind to IPv4 or IPv6.

2018-11-21 11:52:47,099:DEBUG:certbot.error_handler:Calling registered functions
2018-11-21 11:52:47,100:INFO:certbot.auth_handler:Cleaning up challenges
2018-11-21 11:52:47,100:WARNING:certbot.renewal:Attempting to renew cert (www.saarlaendisches-krankenhaus-verzeichnis.de) from /etc/letsencrypt/renewal/www.saarlaendisches-krankenhaus-verzeichnis.de.conf produced an unexpected error: Problem binding to port 443: Could not bind to IPv4 or IPv6.. Skipping.
2018-11-21 11:52:47,102:DEBUG:certbot.renewal:Traceback was:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/certbot/plugins/standalone.py", line 73, in run
    address, self.certs)  # type: acme_standalone.BaseDualNetworkedServers
  File "/usr/lib/python3/dist-packages/acme/standalone.py", line 147, in __init__
    BaseDualNetworkedServers.__init__(self, TLSSNI01Server, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/acme/standalone.py", line 108, in __init__
    raise socket.error("Could not bind to IPv4 or IPv6.")
OSError: Could not bind to IPv4 or IPv6.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/certbot/renewal.py", line 430, in handle_renewal_request
    main.renew_cert(lineage_config, plugins, renewal_candidate)
  File "/usr/lib/python3/dist-packages/certbot/main.py", line 1197, in renew_cert
    renewed_lineage = _get_and_save_cert(le_client, config, lineage=lineage)
  File "/usr/lib/python3/dist-packages/certbot/main.py", line 115, in _get_and_save_cert
    renewal.renew_cert(config, domains, le_client, lineage)
  File "/usr/lib/python3/dist-packages/certbot/renewal.py", line 305, in renew_cert
    new_cert, new_chain, new_key, _ = le_client.obtain_certificate(domains, new_key)
  File "/usr/lib/python3/dist-packages/certbot/client.py", line 334, in obtain_certificate
    orderr = self._get_order_and_authorizations(csr.data, self.config.allow_subset_of_names)
  File "/usr/lib/python3/dist-packages/certbot/client.py", line 370, in _get_order_and_authorizations
    authzr = self.auth_handler.handle_authorizations(orderr, best_effort)
  File "/usr/lib/python3/dist-packages/certbot/auth_handler.py", line 75, in handle_authorizations
    resp = self._solve_challenges(aauthzrs)
  File "/usr/lib/python3/dist-packages/certbot/auth_handler.py", line 126, in _solve_challenges
    resp = self.auth.perform(all_achalls)
  File "/usr/lib/python3/dist-packages/certbot/plugins/standalone.py", line 234, in perform
    return [self._try_perform_single(achall) for achall in achalls]
  File "/usr/lib/python3/dist-packages/certbot/plugins/standalone.py", line 234, in <listcomp>
    return [self._try_perform_single(achall) for achall in achalls]
  File "/usr/lib/python3/dist-packages/certbot/plugins/standalone.py", line 241, in _try_perform_single
    _handle_perform_error(error)
  File "/usr/lib/python3/dist-packages/certbot/plugins/standalone.py", line 239, in _try_perform_single
    return self._perform_single(achall)
  File "/usr/lib/python3/dist-packages/certbot/plugins/standalone.py", line 247, in _perform_single
    servers, response = self._perform_tls_sni_01(achall)
  File "/usr/lib/python3/dist-packages/certbot/plugins/standalone.py", line 264, in _perform_tls_sni_01
    servers = self.servers.run(port, challenges.TLSSNI01, listenaddr=addr)
  File "/usr/lib/python3/dist-packages/certbot/plugins/standalone.py", line 78, in run
    raise errors.StandaloneBindError(error, port)
certbot.errors.StandaloneBindError: Problem binding to port 443: Could not bind to IPv4 or IPv6.
```

## Here is the relevant nginx server block or Apache virtualhost for the domain I am configuring:
Not relevant

## Here is the relevant renewal config for the domain I am renewing:
```
# renew_before_expiry = 30 days
version = 0.26.1
archive_dir = /etc/letsencrypt/archive/www.saarlaendisches-krankenhaus-verzeichnis.de
cert = /etc/letsencrypt/live/www.saarlaendisches-krankenhaus-verzeichnis.de/cert.pem
privkey = /etc/letsencrypt/live/www.saarlaendisches-krankenhaus-verzeichnis.de/privkey.pem
chain = /etc/letsencrypt/live/www.saarlaendisches-krankenhaus-verzeichnis.de/chain.pem
fullchain = /etc/letsencrypt/live/www.saarlaendisches-krankenhaus-verzeichnis.de/fullchain.pem

# Options used in the renewal process
[renewalparams]
server = https://acme-v02.api.letsencrypt.org/directory
account = <redacted>
http01_port = 54321
authenticator = standalone
```

**Comment by fdellwing** [comment]
FYI: The cert got first deployed with this command: `sudo certbot certonly --standalone --http-01-port 54321 -d www.saarlaendisches-krankenhaus-verzeichnis.de -d saarlaendisches-krankenhaus-verzeichnis.de -d krsaar.live.promato.de`

**Comment by fdellwing** [comment]
It works after adding `pref_challs = http-01,` to the renewal conf.

The main question still stands:

Why can I not renew a certificate automaticly after getting it with this command?

**Comment by schoen** [comment]
Hi @fdellwing,

The answer is that the default in all versions of Certbot until recently was to prefer TLS-SNI-01 challenges over HTTP-01 challenges. But the option `--http-01-port 54321` only changes the port that Certbot uses for HTTP-01 challenge, which is not a challenge that was being performed at all (because a TLS-SNI-01 challenge was being performed instead). There is a separate option to change the TLS-SNI-01 port, called `--tls-sni-01-port`. Since you didn't specify a value for `--tls-sni-01-port`, the standalone authenticator used the default of 443 when performing a TLS-SNI-01 challenge.

In general, Certbot doesn't generate a warning if an option affects a code path that is not used (for example, setting an option for an authentication method that isn't the one that you're using to obtain the certificate).

Your change with `pref_challs` is the correct solution to this problem. For other reasons (the TLS-SNI-01 deprecation), this has also become the default in the most recent versions of Certbot:

#6461

**Comment by fdellwing** [comment]
> But the option --http-01-port 54321 only changes the port that Certbot uses for HTTP-01 challenge, which is not a challenge that was being performed at all (because a TLS-SNI-01 challenge was being performed instead).

That is all nice, but after I ran `sudo certbot certonly --standalone --http-01-port 54321 -d ...` it used HTTP-01 challenge (TLS-SNI-01 will always fail, because haproxy binds that port) to get the cert, but it did not setup the renew properly. So there is some inconsitency there between obtaining the first cert and the renew options it will setup.

**Comment by schoen** [comment]
> So there is some inconsitency there between obtaining the first cert and the renew options it will setup.

This is definitely true, and arises because of an awkward issue in the gradual deprecation of the TLS-SNI-01 challenge. In particular, when you first obtain a certificate the CA only offers DNS-01 and HTTP-01 challenges, but when you renew the certificate the CA offers DNS-01, HTTP-01, and TLS-SNI-01 challenges. This difference is implemented at the Let's Encrypt CA as a policy matter because of hypothetical attacks against TLS-SNI-01 in some hosting providers' environments.

Therefore, the `--http-01-port 54321` has a more significant effect for initial issuance with `--standalone` than for renewals, because the renewals (up until the most recent Certbot releases) would default to preferring TLS-SNI-01 even though the initial issuance could not use that challenge method.

As I mentioned, this inconsistency should be eliminated by https://github.com/certbot/certbot/pull/6461, which basically has the same result as your use of `--preferred-challenges http`, but without requiring you to explicitly use this option. That is, the strange behavior that you experienced shouldn't be present for other users in the future because of https://github.com/certbot/certbot/pull/6461, which was merged into the Certbot codebase at the end of October.

I agree that the behavior you experienced from Certbot is quite confusing and annoying, but it still appears to me (1) you've successfully fixed it correctly in the intended way on your end, (2) the reasons for the strange behavior are already understood on our end, and (3) we've also already changed Certbot so that other users who use updated versions in the future will typically not have the same confusing experience. (This change is part of Certbot 0.28 and higher; although your operating system doesn't have such a recent version, eventually more and more of the Certbot-using population should gain access to it.)
